File: 011-020/017.js
//1000までの数の文字列の合計を求めるプログラム
var input = 1000;

// three hundred and forty-two => 23
// one hundred and fifteen => 20

// one, two, three, four, five, six, seven, eight, nine
// ゼロの時はカウントしない
var underTen = [0, 3, 3, 5, 4, 4, 3, 5, 5, 4];

//ten, eleven, twelve, thirteen, fourteen, fifteen, sixteen, seventeen, eighteen, nineteen
var underTwelve = [3, 6, 6, 8, 8, 7, 7, 9, 8, 8];

// ten, twenty, thirty, forty, fifty, sixty, seventy, eighty, ninety
var underHundred = [3, 6, 6, 5, 5, 5, 7, 6, 6];

var hundred = 7;
var thousand = 8;

// 1から1000まで文字数を調べる
// 1000よりも大きい数字は調べられない仕様
var strSum = 0;
for (var i = 1; i <= input; i++) {
  // one thousandの文字数を足す
  if (i === 1000) {
    strSum += 11;
    break;
  }
  // 桁ごとの操作がしやすいように文字列に変換
  var str = i.toString();
  var leng = str.length;
  // 下2桁を取り出す
  var dig2 = parseInt(str.substr(-2, 2), 10);
  if (i >= 100) {
    strSum += underTen[parseInt(str[0], 10)] + hundred;
    // １桁目、２桁目のどちらかが存在するとandが付く
    if (dig2 != 0) {
      strSum += 3;
    }
  }
  if (dig2 >= 10) {
    if (dig2 <= 19) {
      // 10から19までの数字は数え方が特殊
      strSum += underTwelve[parseInt(str.substr(-1, 1), 10)];
      continue;
    } else {
      strSum += underHundred[parseInt(str.substr(-2, 1), 10) - 1];
    }
  }
  strSum += underTen[parseInt(str.substr(-1, 1), 10)];
}


console.log(strSum);

File: 011-020/017.py
# 1000までの数の文字列の合計を求めるプログラム
max_num = 1000

#  three hundred and forty-two => 23
#  one hundred and fifteen => 20

#  one, two, three, four, five, six, seven, eight, nine
#  ゼロの時はカウントしない
under_ten = [0, 3, 3, 5, 4, 4, 3, 5, 5, 4]

# ten, eleven, twelve, thirteen, fourteen, fifteen,
# sixteen, seventeen, eighteen, nineteen
under_twelve = [3, 6, 6, 8, 8, 7, 7, 9, 8, 8]

#  ten, twenty, thirty, forty, fifty, sixty, seventy, eighty, ninety
under_hundred = [3, 6, 6, 5, 5, 5, 7, 6, 6]

hundred = 7

#  1から1000まで文字数を調べる
#  1000よりも大きい数字は調べられない仕様
str_sum = 0
for i in range(1, max_num + 1):
    #  one thousandの文字数を足す
    if i == 1000:
        str_sum += 11
        break
    #  桁ごとの操作がしやすいように文字列に変換
    i_str = str(i)
    #  下2桁を取り出す
    dig2 = int(i_str[-2:])
    if i >= 100:
        str_sum += under_ten[int(i_str[0])] + hundred
        #  １桁目、２桁目のどちらかが存在するとandが付く
        if dig2 != 0:
            str_sum += 3

    if dig2 >= 10:
        if dig2 <= 19:
            #  10から19までの数字は数え方が特殊
            str_sum += under_twelve[int(i_str[-1])]
            continue
        else:
            str_sum += under_hundred[int(i_str[-2]) - 1]
    str_sum += under_ten[int(i_str[-1])]


print(str_sum)
